### PR TITLE
Cl_text_entities_size

### PR DESCRIPTION
**Pull request by BannZay**
Allows to users to change size of overlay text entities

### CODE REVIEW

**Comment by Jupeyy** [comment]
Could you please undo all changes in CMapImages.cpp.

It was on purpose designed to fit the text overlay into 1 tile, to drastically increase performance for all kind of overlay that uses text, when you use OpenGL 3.3

The text is already aligned to fit the tile

Also notice the FPS difference!!

OpenGL 3.3

screenshot_2019-04-13_08-52-41

OpenGL 2

screenshot_2019-04-13_08-52-57

**Comment by fokkonaut** [comment]
I am against this change

**Comment by BannZay** [comment]
Just tested output for both 2.0 and 3.3 OpenGL before and after my commit. Still cant get your point, @Jupeyy  
Whats wrong?
There IS differences but they are pretty minor
OpenGL 3.3:
Before commit:
3 3_before
After:
3 3_after

OpenGL 2.0
Before:
2 0_before
After:
2 0_after

> The text is already aligned to fit the tile
What if someone has a 27 inches monitor where text looks ugly being so big as tile?
There is cl_text_entities_size coming in

**Comment by Jupeyy** [comment]
because your merge conflict redone the old behaivour, not because your code is working

**Comment by BannZay** [comment]
> because your merge conflict redone the old behaivour, not because your code is working

Damn, failed to merge. I will fix soon

**Comment by fokkonaut** [comment]
> What if someone has a 27 inches monitor where text looks ugly being so big as tile?
> There is cl_text_entities_size coming in

if someone has a screen that big, he would use zoom- multiple times (i use it once and i have a 22 inch monitor) so the tiles are smaller again.

**Comment by BannZay** [comment]
I just want to have a solution for my problem, and zooming out didnt helped me. 
And there I am.
Anyway, whats bad in possibility to configure text size, especially if user wants this.

**Comment by fokkonaut** [comment]
Because its automatically set to fit the box of the tile

**Comment by BannZay** [comment]
> Because its automatically set to fit the box of the tile

No changes at all with default value for OpenGL 3.3. Check out screenshots

File: src/game/client/render_map.cpp [hunk 1]
```diff
@@ -415,6 +415,8 @@ void CRenderTools::RenderTeleOverlay(CTeleTile *pTele, int w, int h, float Scale
 	if(EndX - StartX > Graphics()->ScreenWidth() / g_Config.m_GfxTextOverlay || EndY - StartY > Graphics()->ScreenHeight() / g_Config.m_GfxTextOverlay)
 		return; // its useless to render text at this distance
 
+	float size = g_Config.m_ClTextEntitiesSize / 64.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float Size = g_Config.m_ClTextEntitiesSize / 64.f;
```
Style

File: src/game/client/render_map.cpp [hunk 2]
```diff
@@ -439,7 +441,7 @@ void CRenderTools::RenderTeleOverlay(CTeleTile *pTele, int w, int h, float Scale
 				char aBuf[16];
 				str_format(aBuf, sizeof(aBuf), "%d", Index);
 				UI()->TextRender()->TextColor(1.0f, 1.0f, 1.0f, Alpha);
-				UI()->TextRender()->Text(0, mx*Scale - 3.f, my*Scale, Scale - 5.f, aBuf, -1);
+				UI()->TextRender()->Text(0, mx*Scale - 3.f, my*Scale, size * Scale, aBuf, -1);
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
				UI()->TextRender()->Text(0, mx*Scale - 3.f, my*Scale, Size * Scale, aBuf, -1);
```
^^

File: src/game/client/render_map.cpp [hunk 3]
```diff
@@ -460,6 +462,8 @@ void CRenderTools::RenderSpeedupOverlay(CSpeedupTile *pSpeedup, int w, int h, fl
 	if(EndX - StartX > Graphics()->ScreenWidth() / g_Config.m_GfxTextOverlay || EndY - StartY > Graphics()->ScreenHeight() / g_Config.m_GfxTextOverlay)
 		return; // its useless to render text at this distance
 
+	float size = g_Config.m_ClTextEntitiesSize / 64.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float Size = g_Config.m_ClTextEntitiesSize / 64.f;
```
Style

File: src/game/client/render_map.cpp [hunk 4]
```diff
@@ -529,6 +533,8 @@ void CRenderTools::RenderSwitchOverlay(CSwitchTile *pSwitch, int w, int h, float
...
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float Size = g_Config.m_ClTextEntitiesSize / 64.f;
```
Style

File: src/game/client/render_map.cpp [hunk 5]
```diff
@@ -587,6 +593,8 @@ void CRenderTools::RenderTuneOverlay(CTuneTile *pTune, int w, int h, float Scale
 	if(EndX - StartX > Graphics()->ScreenWidth() / g_Config.m_GfxTextOverlay || EndY - StartY > Graphics()->ScreenHeight() / g_Config.m_GfxTextOverlay)
 		return; // its useless to render text at this distance
 
+	float size = g_Config.m_ClTextEntitiesSize * Scale / 64.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float Size = g_Config.m_ClTextEntitiesSize * Scale / 64.f;
```
Style

File: src/game/client/components/mapimages.cpp [hunk 6]
```diff
@@ -196,3 +147,89 @@ int CMapImages::GetOverlayCenter()
 {
 	return m_OverlayCenterTexture;
 }
+
+void CMapImages::TextureSize(int Size)
+{
+	if (m_TextureSize == Size)
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	if(m_TextureSize == Size)
```
Style

File: src/game/client/gameclient.cpp [hunk 7]
```diff
@@ -2042,6 +2045,22 @@ void CGameClient::ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserDa
 			g_Config.m_ClDummy = 0;
 }
 
+void CGameClient::ConchainClTextEntitiesSize(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData)
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]: Clean this up, should take about 3 lines, 5 if you are being strict with style.

File: src/game/client/gameclient.cpp [hunk 8]
```diff
@@ -2042,6 +2045,22 @@ void CGameClient::ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserDa
...
+{
+	pfnCallback(pResult, pCallbackUserData);
+
+	if (pResult->NumArguments())
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]:
```suggestion
	if(pResult->NumArguments())
```
Style

File: src/game/client/components/mapimages.cpp [hunk 9]
```diff
@@ -196,3 +147,89 @@ int CMapImages::GetOverlayCenter()
...
+		return;
+
+	m_TextureSize = Size;
+
+	if (Graphics()) // check if component was initialized
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	if(Graphics()) // check if component was initialized
```
Style

I think you should/could also check `m_OverlayBottomTexture == -1`

File: src/game/client/components/mapimages.cpp [hunk 10]
```diff
@@ -196,3 +147,89 @@ int CMapImages::GetOverlayCenter()
...
+		Graphics()->UnloadTexture(m_OverlayTopTexture);
+		Graphics()->UnloadTexture(m_OverlayCenterTexture);
+
+		m_OverlayBottomTexture = m_OverlayTopTexture = m_OverlayCenterTexture = -1;
+
+		InitOverlayTextures();
+	}
+}
+
+int CMapImages::TextureSize()
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]: This could be in the header file

File: src/game/client/components/mapimages.h [hunk 11]
```diff
@@ -23,19 +24,24 @@ class CMapImages : public CComponent
...
 	// DDRace
-
 	int GetEntities();
 	
 	int GetOverlayBottom();
 	int GetOverlayTop();
 	int GetOverlayCenter();
 
+	void TextureSize(int Size);
+	int TextureSize();
```
Comment (Learath2) [inline src/game/client/components/mapimages.h]: I'm not really sure if it's a good idea to name these as the same, but not really important.

**Review by Learath2 — CHANGES_REQUESTED** [review]
Looks good except for a couple style issues.

Also maybe we should make this setting between 0 and 100, 0 and 64 isn't really intuitive, but again, not really important.

**Comment by 12pm** [comment]
probably better to have the config in %?

File: src/game/client/gameclient.cpp (hunk 7 above)
Comment (BannZay) [reply src/game/client/gameclient.cpp]: Just followed code around style it will looks unnatural if it will be done

**Comment by BannZay** [comment]
Fixed code styles issues. Completely agree setting should be as percent so made it as percent also

File: src/game/client/components/mapimages.cpp (hunk 10 above)
Comment (BannZay) [reply src/game/client/components/mapimages.cpp]: So getter will be in header and setter not ? If it is okay I can do it.

File: src/game/client/gameclient.cpp (hunk 7 above)
Comment (Learath2) [reply src/game/client/gameclient.cpp]: I meant the entire function, it's all messy, has an unneeded block there, a commented out if statement.

File: src/game/client/components/mapimages.cpp (hunk 10 above)
Comment (Learath2) [reply src/game/client/components/mapimages.cpp]: I'd call the setter `SetTextureSize` and move the getter to the header

File: src/game/client/gameclient.cpp (hunk 7 above)
Comment (BannZay) [reply src/game/client/gameclient.cpp]: I see, my bad, sorry. Didnt realized that I changed code before reading this comment. Anyway, it should looks good now

File: src/game/client/components/mapimages.cpp (hunk 10 above)
Comment (BannZay) [reply src/game/client/components/mapimages.cpp]: I am worried about that many commits already here. I will do it when something else is need to be changed

**Comment by fokkonaut** [comment]
So the default value is 100 and it cant get bigger than it is right now, correct? If yes the rest looks done.

**Comment by BannZay** [comment]
@fokkonaut, yes. 100 - is default and maximum value for this setting. With 100 it will looks the same as now

**Comment by fokkonaut** [comment]
Okay, good. This looks done

**Comment by VxMiMxV** [comment]
Just wanted to say that i'd love to have this.
Also:
Screen 1698
I hope the single digit numbers will be centered like this (4 is unedited)
And yes, some numbers stick out like this 4 for me personally.

**Comment by BannZay** [comment]
@VxMiMxV, I believe you`ve posted screenshot with modified tileset (btw, I do use similar). With original one the number will overlap with text "FROM" so there is some questions like "should we lift it up to prevent overlapping?". It is responsibility for another setting like "overlay_text_centered or better overlay_text_x/overlay_text_y" tho.

This one setting is first stem of such custumization.

 I would love to continue working and discussing it with support of people who could make such decisions. Because doing big change may open such many questions so this PR will never succeed

**Comment by def-** [comment]
@Jupeyy Is this ok from your side now?

**Comment by Jupeyy** [comment]
Mh not 100% happy tbh.
The speedups text is now always 50%(at 100% setting) of the tile, which is pretty huge, the value before was 20
so 20 / 64 around 31% of the tile

also the y offset for the bottom overlay isn't calculated properly if you use smaller value. it has to stick to the bottom of the tile not to the mid of the tile.

And last, what @Learath2 said, please name the getter and setter properly, else the code gets just more messy to read

**Comment by BannZay** [comment]
> speedups text is now always 50%

I just dropped all the magic instead of figuring out what the hell is going on here. Once I realized it is fine I just left it unless I dont have any requirements how this should work.

P.s.: it will be nice if some one could rewrite whole functionality from the scratch and this PR will be turned to POC

**Comment by Jupeyy** [comment]
its totally fine and useful commit, pls don't give up:

imagine the whole mapimages thing like this. all the game tile attributes have a range of 0-255(like speed, or tele outcome)
and teeworlds map layers are also 256 possiblities, that made it easy to overlap tiles with the text.
the magic is just that it pre renders all the text into textures, so at runtime it doesnt need CPU nor GPU workload.

It must fit 1 tile, because, then it can be querried just like a game tile would.

Here are images, that probably explain what happens

image

image

image


You do everything right, i just thought would be neat to actually fit stuff correctly, a tile on these textures is 64 pixel, so if you use a font size of 20, you could just do 64 - 20, that would make the text appear on the bottom of the tile.

Also nothing is wrong with many commits as you said somewhere. it's totally fine :)

**Comment by BannZay** [comment]
I will fix everything asked above. Just one question - it will be fine if text vertically aligned to the center of it 32 pixel area?

**Comment by Jupeyy** [comment]
well alignment is something that could be improved for sure(also horizontal, bcs small letters are on the left not centered anymore), but i guess its ok for now, since the standard value atleast looks good then.

But if you want to work on this sure, go ahead ;D

**Comment by BannZay** [comment]
**Before:**
before
**Now:**
now

**Comment by Jupeyy** [comment]
i just let the mantainers decide, if they are ok, its ok

File: src/engine/client/text.cpp [hunk 12]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
-				FontSize = MaxSize;
-
-			pCurrent = (char *)pText;
-			pEnd = pCurrent + Length;
-		}
-
-
 		while(pCurrent < pEnd)
-		{
+		{
```
Comment (Learath2) [inline src/engine/client/text.cpp]: Could you please cleanup the trailing spaces?

File: src/engine/client/text.cpp [hunk 13]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
-					WidthLastChars += (SlotW + 1);
+		if (MaxSize > 0 && FontSize > MaxSize)
+			FontSize = MaxSize;
+
+		return FontSize;
+	}
+
+	virtual int CalculateTextWidth(const char* pText, int TextLength, int FontWidth, int FontHeight)
+	{
+		CFont* pFont = m_pDefaultFont;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		CFont *pFont = m_pDefaultFont;
```
We put the * with the label.

File: src/engine/client/text.cpp [hunk 14]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+		const char* pEnd = pCurrent + TextLength;
+
+		int WidthOfText = 0;
+		FT_Set_Pixel_Sizes(pFont->m_FtFace, FontWidth, FontHeight);
+		while (pCurrent < pEnd)
+		{
+			const char* pTmp = pCurrent;
+			int NextCharacter = str_utf8_decode(&pTmp);
+
+			if (NextCharacter)
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
			if(NextCharacter)
```

No space after control keywords

File: src/engine/client/text.cpp [hunk 15]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+			{
+				FT_Int32 FTFlags = 0;
+#if FREETYPE_MAJOR >= 2 && FREETYPE_MINOR >= 7 && (FREETYPE_MINOR > 7 || FREETYPE_PATCH >= 1)
+				FTFlags = FT_LOAD_BITMAP_METRICS_ONLY | FT_LOAD_NO_BITMAP;
+#else
+				FTFlags = FT_LOAD_RENDER | FT_LOAD_NO_BITMAP;
+#endif
+				if (FT_Load_Char(pFont->m_FtFace, NextCharacter, FTFlags))
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
				if(FT_Load_Char(pFont->m_FtFace, NextCharacter, FTFlags))
```
^^

File: src/engine/textrender.h [hunk 16]
```diff
@@ -98,7 +98,8 @@ class ITextRender : public IInterface
 	virtual void RenderTextContainer(int TextContainerIndex, STextRenderColor *pTextColor, STextRenderColor *pTextOutlineColor) = 0;
 	virtual void RenderTextContainer(int TextContainerIndex, STextRenderColor *pTextColor, STextRenderColor *pTextOutlineColor, float X, float Y) = 0;
 
-	virtual void UploadEntityLayerText(int TextureID, const char *pText, int Length, float x, float y, int Size, int MaxWidth, int MaxSize = -1, int MinSize = -1) = 0;
+	virtual void UploadEntityLayerText(int TextureID, const char* pText, int Length, float x, float y, int FontHeight) = 0;
```
Comment (Learath2) [inline src/engine/textrender.h]:
```suggestion
	virtual void UploadEntityLayerText(int TextureID, const char *pText, int Length, float x, float y, int FontHeight) = 0;
```
^^

File: src/game/client/components/mapimages.cpp [hunk 17]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
 {
 	return m_OverlayCenterTexture;
 }
+
+void CMapImages::SetTextureScale(int Scale)
+{
+	if (m_TextureScale == Scale)
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	if(m_TextureScale == Scale)
```
^^

File: src/game/client/components/mapimages.cpp [hunk 18]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+}
+
+int CMapImages::GetTextureScale()
+{
+	return m_TextureScale;
+}
+
+int CMapImages::UploadEntityLayerText(int TextureSize, int YOffset)
+{	
+	void* pMem = calloc(1024 * 1024, 1);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	void *pMem = calloc(1024 * 1024, 1);
```
^^

File: src/game/client/components/mapimages.cpp [hunk 19]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+	free(pMem);
+
+	char buff[4];
+	str_format(buff, 4, "%d", 255);
+	int len = str_length(buff);
+
+	int FontSize = TextRender()->AdjustFontSize(buff, len, TextureSize);
+	YOffset += ((TextureSize - FontSize)/2);
+
+	for (int i = 0; i < 256; ++i)
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	for(int i = 0; i < 256; ++i)
```
^^

File: src/game/client/components/mapimages.cpp [hunk 20]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+	return m_TextureScale;
+}
+
+int CMapImages::UploadEntityLayerText(int TextureSize, int YOffset)
+{	
+	void* pMem = calloc(1024 * 1024, 1);
+	int TextureID = Graphics()->LoadTextureRaw(1024, 1024, CImageInfo::FORMAT_ALPHA, pMem, CImageInfo::FORMAT_ALPHA, IGraphics::TEXLOAD_NOMIPMAPS);
+	free(pMem);
+
+	char buff[4];
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	char aBuf[4];
```
Naming

File: src/game/client/components/mapimages.cpp [hunk 21]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+	str_format(buff, 4, "%d", 255);
+	int len = str_length(buff);
+
+	int FontSize = TextRender()->AdjustFontSize(buff, len, TextureSize);
+	YOffset += ((TextureSize - FontSize)/2);
+
+	for (int i = 0; i < 256; ++i)
+	{
+		str_format(buff, 4, "%d", i);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
		int Len = str_format(aBuf, 4, "%d", i);
```
^^ VV

File: src/game/client/components/mapimages.cpp [hunk 22]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+		int len = str_length(buff);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion

```
`str_format` already returns the length of the string written.

File: src/game/client/components/mapimages.cpp [hunk 23]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+}
+
+int CMapImages::UploadEntityLayerText(int TextureSize, int YOffset)
+{	
+	void* pMem = calloc(1024 * 1024, 1);
+	int TextureID = Graphics()->LoadTextureRaw(1024, 1024, CImageInfo::FORMAT_ALPHA, pMem, CImageInfo::FORMAT_ALPHA, IGraphics::TEXLOAD_NOMIPMAPS);
+	free(pMem);
+
+	char buff[4];
+	str_format(buff, 4, "%d", 255);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
	int Len = str_format(aBuf, 4, "%d", 255);
```
^^VV

File: src/game/client/components/mapimages.cpp [hunk 24]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+	int len = str_length(buff);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion

```
VV

File: src/game/client/components/mapimages.cpp [hunk 25]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+	if(m_OverlayTopTexture == -1)
+	{
+		m_OverlayTopTexture = UploadEntityLayerText(TextureSize/2, TextureToVerticalCenterOffset/2);
+	}
+
+	if(m_OverlayCenterTexture == -1)
+	{
+		m_OverlayCenterTexture = UploadEntityLayerText(TextureSize, TextureToVerticalCenterOffset);
+	}
+}
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
}

```
You should configure your editor to make sure it adds a newline at EOF.

File: src/game/client/render_map.cpp [hunk 26]
```diff
@@ -415,6 +415,9 @@ void CRenderTools::RenderTeleOverlay(CTeleTile *pTele, int w, int h, float Scale
 	if(EndX - StartX > Graphics()->ScreenWidth() / g_Config.m_GfxTextOverlay || EndY - StartY > Graphics()->ScreenHeight() / g_Config.m_GfxTextOverlay)
 		return; // its useless to render text at this distance
 
+	float size = g_Config.m_ClTextEntitiesSize/100.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float Size = g_Config.m_ClTextEntitiesSize/100.f;
```
Naming

File: src/game/client/components/mapimages.cpp [hunk 27]
```diff
@@ -196,3 +147,76 @@ int CMapImages::GetOverlayCenter()
...
+
+	for (int i = 0; i < 256; ++i)
+	{
+		str_format(buff, 4, "%d", i);
+		int len = str_length(buff);
+
+		float x = (i%16)*64;
+		float y = (i/16)*64;
+		
+		TextRender()->UploadEntityLayerText(TextureID, buff, len, x, y + YOffset, FontSize);
```
Comment (Learath2) [inline src/game/client/components/mapimages.cpp]:
```suggestion
		TextRender()->UploadEntityLayerText(TextureID, aBuf, Len, x, y + YOffset, FontSize);
```
^^

File: src/game/client/render_map.cpp [hunk 28]
```diff
@@ -415,6 +415,9 @@ void CRenderTools::RenderTeleOverlay(CTeleTile *pTele, int w, int h, float Scale
...
+	float toCenterOffset = (1-size)/2.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
	float ToCenterOffset = (1-size)/2.f;
```
^^

File: src/game/client/render_map.cpp [hunk 29]
```diff
@@ -439,7 +442,7 @@ void CRenderTools::RenderTeleOverlay(CTeleTile *pTele, int w, int h, float Scale
 				char aBuf[16];
 				str_format(aBuf, sizeof(aBuf), "%d", Index);
 				UI()->TextRender()->TextColor(1.0f, 1.0f, 1.0f, Alpha);
-				UI()->TextRender()->Text(0, mx*Scale - 3.f, my*Scale, Scale - 5.f, aBuf, -1);
+				UI()->TextRender()->Text(0, mx*Scale - 3.f, (my+toCenterOffset)*Scale, size*Scale, aBuf, -1);
```
Comment (Learath2) [inline src/game/client/render_map.cpp]:
```suggestion
				UI()->TextRender()->Text(0, mx*Scale - 3.f, (my+ToCenterOffset)*Scale, Size*Scale, aBuf, -1);
```
^^

File: src/game/client/render_map.cpp [hunk 30]
```diff
@@ -460,6 +463,9 @@ void CRenderTools::RenderSpeedupOverlay(CSpeedupTile *pSpeedup, int w, int h, fl
 	if(EndX - StartX > Graphics()->ScreenWidth() / g_Config.m_GfxTextOverlay || EndY - StartY > Graphics()->ScreenHeight() / g_Config.m_GfxTextOverlay)
 		return; // its useless to render text at this distance
 
+	float size = g_Config.m_ClTextEntitiesSize / 100.f;
```
Comment (Learath2) [inline src/game/client/render_map.cpp]: Naming as above, also for lines below, too many to mark :P

**Review by Learath2 — CHANGES_REQUESTED** [review]
A lot of style nits, other then that, if we agree this is the best way to size them then it LGTM afterwards.

File: src/engine/client/text.cpp [hunk 31]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
-			return;
-
-		// set length
-		if(Length < 0)
-			Length = str_length(pText);
-
-		const char *pCurrent = (char *)pText;
-		const char *pEnd = pCurrent+Length;
-
+		const char* pCurrent = (char*)pText;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		const char  pCurrent = (char *)pText;
```

File: src/engine/client/text.cpp [hunk 32]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
+		const char* pEnd = pCurrent + Length;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		const char *pEnd = pCurrent + Length;
```

File: src/engine/client/text.cpp [hunk 33]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
+		CFont* pFont = m_pDefaultFont;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		CFont *pFont = m_pDefaultFont;
```

File: src/engine/client/text.cpp [hunk 34]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
+		FT_Bitmap* pBitmap;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		FT_Bitmap *pBitmap;
```

File: src/engine/client/text.cpp [hunk 35]
```diff
@@ -1698,69 +1698,19 @@ class CTextRender : public IEngineTextRender
...
-			pEnd = pCurrent + Length;
-		}
-
-
 		while(pCurrent < pEnd)
 		{
 			const char *pTmp = pCurrent;
 			int NextCharacter = str_utf8_decode(&pTmp);
-
+
```
Comment (Learath2) [inline src/engine/client/text.cpp]: Can't suggest this one but trailing spaces

File: src/engine/client/text.cpp [hunk 36]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+				
+				Graphics()->LoadTextureRawSub(TextureID, x + WidthLastChars, y, SlotW, SlotH, CImageInfo::FORMAT_ALPHA, ms_aGlyphData);
+				WidthLastChars += (SlotW + 1);
+				
+			}
+			pCurrent = pTmp;
+		}
+	}
+	
+	virtual int AdjustFontSize(const char* pText, int TextLength, int MaxSize = -1)
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
	virtual int AdjustFontSize(const char *pText, int TextLength, int MaxSize = -1)
```

File: src/engine/client/text.cpp [hunk 37]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+
+		int FontSize = 100.f / ((float)WidthOfText / (float)MaxSize);
+
+		if (MaxSize > 0 && FontSize > MaxSize)
+			FontSize = MaxSize;
+
+		return FontSize;
+	}
+
+	virtual int CalculateTextWidth(const char* pText, int TextLength, int FontWidth, int FontHeight)
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
	virtual int CalculateTextWidth(const char *pText, int TextLength, int FontWidth, int FontHeight)
```

File: src/engine/client/text.cpp [hunk 38]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+	{
+		CFont *pFont = m_pDefaultFont;
+		const char* pCurrent = (char*)pText;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		const char *pCurrent = (char *)pText;
```

File: src/engine/client/text.cpp [hunk 39]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+		const char* pEnd = pCurrent + TextLength;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
		const char *pEnd = pCurrent + TextLength;
```

File: src/engine/client/text.cpp [hunk 40]
```diff
@@ -1775,34 +1725,79 @@ class CTextRender : public IEngineTextRender
...
+
+		int WidthOfText = 0;
+		FT_Set_Pixel_Sizes(pFont->m_FtFace, FontWidth, FontHeight);
+		while (pCurrent < pEnd)
+		{
+			const char* pTmp = pCurrent;
```
Comment (Learath2) [inline src/engine/client/text.cpp]:
```suggestion
			const char *pTmp = pCurrent;
```

File: src/engine/textrender.h [hunk 41]
```diff
@@ -98,7 +98,8 @@ class ITextRender : public IInterface
 	virtual void RenderTextContainer(int TextContainerIndex, STextRenderColor *pTextColor, STextRenderColor *pTextOutlineColor) = 0;
 	virtual void RenderTextContainer(int TextContainerIndex, STextRenderColor *pTextColor, STextRenderColor *pTextOutlineColor, float X, float Y) = 0;
 
-	virtual void UploadEntityLayerText(int TextureID, const char *pText, int Length, float x, float y, int Size, int MaxWidth, int MaxSize = -1, int MinSize = -1) = 0;
+	virtual void UploadEntityLayerText(int TextureID, const char *pText, int Length, float x, float y, int FontHeight) = 0;
+	virtual int AdjustFontSize(const char* pText, int TextLength, int MaxSize = -1) = 0;
```
Comment (Learath2) [inline src/engine/textrender.h]:
```suggestion
	virtual int AdjustFontSize(const char *pText, int TextLength, int MaxSize = -1) = 0;
```

File: src/game/client/gameclient.cpp [hunk 42]
```diff
@@ -1787,6 +1790,18 @@ void CGameClient::ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserDa
 			g_Config.m_ClDummy = 0;
 }
 
+void CGameClient::ConchainClTextEntitiesSize(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData)
+{
+	pfnCallback(pResult, pCallbackUserData);
+
+	if(pResult->NumArguments())
+	{
+		CGameClient* GameClient = (CGameClient*)pUserData;
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]:
```suggestion
		CGameClient *pGameClient = (CGameClient*)pUserData;
```

File: src/game/client/gameclient.cpp [hunk 43]
```diff
@@ -1787,6 +1790,18 @@ void CGameClient::ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserDa
...
+		GameClient->m_pMapimages->SetTextureScale(g_Config.m_ClTextEntitiesSize);
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]:
```suggestion
		pGameClient->m_pMapimages->SetTextureScale(g_Config.m_ClTextEntitiesSize);
```

File: src/game/client/gameclient.cpp [hunk 44]
```diff
@@ -1787,6 +1790,18 @@ void CGameClient::ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserDa
...
+	}
+}
+
+
```
Comment (Learath2) [inline src/game/client/gameclient.cpp]:
```suggestion

```

File: src/game/client/gameclient.h [hunk 45]
```diff
@@ -118,6 +118,7 @@ class CGameClient : public IGameClient
 	static void ConchainSpecialInfoupdate(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData);
 	static void ConchainSpecialDummyInfoupdate(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData);
 	static void ConchainSpecialDummy(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData);
+	static void ConchainClTextEntitiesSize(IConsole::IResult * pResult, void * pUserData, IConsole::FCommandCallback pfnCallback, void * pCallbackUserData);
```
Comment (Learath2) [inline src/game/client/gameclient.h]:
```suggestion
	static void ConchainClTextEntitiesSize(IConsole::IResult *pResult, void *pUserData, IConsole::FCommandCallback pfnCallback, void *pCallbackUserData);
```

**Review by Learath2 — CHANGES_REQUESTED** [review]
There are still more of these, You really should configure your editor.

Also please capitalize `toCenterOffset` properly for all instances.

**Comment by BannZay** [comment]
> You really should configure your editor

I want it to but can not find anything about how to do it. Furtermore, I could not even find DDNet code style convention. For example toCenterOffset is local variable and even in Teeworlds code convention nothing about local variables.

Maybe there is nothing special for you who work with DDNet long enough, but for me, who dont even code in C++ it is challenge. It just not worth it. I have to analyze some amount of code to configure whole code style to contribute a little thing.

I use Visual Studio 2019. There is [thing](https://devblogs.microsoft.com/dotnet/code-style-configuration-in-the-vs2017-rc-update/) what probably can solve it. So newcomers can easily use it

**Comment by BannZay** [comment]
Everything I said before is not excuse, just wanted to let you know about problem I face working on this. And everything you reported I will fix

**Comment by def-** [comment]
@Learath2 can we squash and merge?

**Pull request merged**